### ISSUE TITLE
Your wordpress server appears to be overloaded.

### ISSUE THREAD

**Issue by anicioalexandre** [gatsbyjs/gatsby#29971]
Hi guys!

I'm getting the error below when trying to call `gatsby develop`.

```
 ERROR #gatsby-source-wordpress-experimental_111007 

 gatsby-source-wordpress  Your wordpress server at http://localwp/graphql appears to be overloaded.

Try reducing the requestConcurrency for content updates or the previewRequestConcurrency for previews:

{
  resolve: 'gatsby-source-wordpress-experimental',
  options: {
    schema: {
      requestConcurrency: 5, // currently set to 5
      previewRequestConcurrency: 2, // currently set to 2
    }
  },
}

The GATSBY_CONCURRENT_REQUEST environment variable is no longer used to control concurrency.
If you were previously using that, you'll need to use the settings above instead.
```

I tried to set `requestConcurrency: 5` and `previewRequestConcurrency: 2` as suggested but it was not enough.

Things I did before getting error appeared that may be related:
- I updated the WPGraphQL plugin (v1.1.1) today (I tried to downgrade to v1.1.0 but the problem persists)
- I set `useGatsbyImage` (within the 'gatsby-source-wordpress-experimental' options in gatsby-config.js) to false to do some tests.

Plugins used on Wordpress are: `WP Gatsby (v0.9.1)`,  `WP GraphQL (v1.1.0)`, `WP GraphQL Gutenberg (v0.3.5)`.

I'm running a local wordpress server (via Local) and it seems to be working normally.

Appreciate any help!

**Comment by stevepepple** [comment]
> Our team recently managed to have a quite stable setup with Gatsby / Wordpress / Gutenberg integration using WP Gatsby, WP GraphQL, WP GraphQL Gutenberg. Just sharing our experience if it can help
> 
> ## TL;DR
> Excluding non desired types and fields reduced the size of a **single** GraphQL query made to Wordpress server from 790 Ko to 4 Ko ! This tremendously cut build times and our wordpress server never overloads.
> 
> ## Long story
> Our Gatsby / Wordpress / Gutenberg setup was overloading the server, causing our builds to crash intermittently, and to last at least 30 minutes on a clean build.
> 
> Digging in the GraphQL schema, we realized how many custom types and attributes we were loading in the schema, even though our content writers never used the vast majority of them. Most of them were coming from Gutenberg plugins.
> 
> Using the [`writeQueriesToDisk` ](https://github.com/gatsbyjs/gatsby-source-wordpress-experimental/blob/master/docs/plugin-options.md#debuggraphqlwritequeriestodisk-boolean) parameter, we were able to see just how big were the sizes of the GraphQL queries when the WP GraphQL Gutenberg was enabled : 790 Ko for a Post or for a Category.
> 
> By excluding all the `Types` and all the `Fields` that we don't use, we reduced the size of a single query to 4Ko
> Here is an extract of our config : https://gist.github.com/babgyy/cba9f6f4b5b737913739ff4d9eb034b1
> 
> Now, a clean build lasts ~10 minutes which is not much more than it was before, most of the time being spent dealing with images. And our Wordpress server stopped being overloaded.
> 
> Only downside is that our wordpress users are limited in what they can use, but that forces us to communicate and prevents unexpected results.
> 
> Cutting down the queries bit by bit : ⬇️
> ![image](https://user-images.githubusercontent.com/852926/105159027-e8e62a00-5b0e-11eb-9a1b-882f29283b38.png)

This is a really great insight. Thanks!

**Comment by TylerBarnes** [comment]
@stevepepple thanks for sharing that, that's certainly a very dramatic difference. We've had the idea for quite a while to automatically exclude unused fields and types in production builds but haven't prioritized working on that yet. The data you shared here helps outline why it would be such a compelling feature to add, so thanks for sharing!

**Comment by peach-inzn** [comment]
I am having the same issue with WPGraphQL Gutenberg 0.3.11 and WPGraphQL 1.6.12. Also tried it with 
```
schema: {
          perPage: 20,
          timeout: 300000,
          requestConcurrency: 5,
          previewRequestConcurrency: 2,
}
```
but that doesn't change anything. All kinds of debug logging fails (both WP and GraphQL).

<img width="873" alt="Screenshot 2022-02-28 at 10 37 45 AM" src="https://user-images.githubusercontent.com/100540199/155959901-e4242a5c-9054-409a-92b0-169035f00ba4.png">

**Comment by vimarshacooray** [comment]
> I believe the "server overloaded" error is a misrepresentation of the actual error in many cases. If anyone experiencing this can email me a public `/graphql` endpoint I can use to reproduce that would be helpful :) ([tyler@gatsbyjs.com](mailto:tyler@gatsbyjs.com))
> 
> Thanks everyone

Facing the Same issue when building on actions pipelines. But i have 0 Issues build on local. the error message is misinterpreting as @TylerBarnes  mentioned. my WP server definitely not overloading as Gatsby builds fine on local.

**Comment by TylerBarnes** [comment]
@vimarshacooray in your case it might be accurate because your local internet connection is likely much slower than the connection in your pipeline. Faster internet speed means it will try to fetch data and files more quickly which can cause your WP server to stop responding.

**Comment by vimarshacooray** [comment]
> @vimarshacooray in your case it might be accurate because your local internet connection is likely much slower than the connection in your pipeline. Faster internet speed means it will try to fetch data and files more quickly which can cause your WP server to stop responding.

@TylerBarnes I just found the reason for this. it had nothing to do with my wordpress overloading or internet speed.
its an 8GB 4C server with 5 posts on it with no traffic. no way its going to overload that easily. memory was limited to 128MB but increased it to 4GB on WP. i have dropped it down to 128MB after finding the issue.

-> Cloudflare was blocking the request on the pipeline (bot defend mode). ⚰️😅
but it allowed the connection when running locally.

Hope this info helps someone!

**Comment by TylerBarnes** [comment]
@vimarshacooray glad you figured it out! 😄

**Comment by Harry-Gardiner** [comment]
Ended up on this thread trying to configure a headless WordPress site with Gatsby - Dec 2022. 

I've managed to get it working with the required plugins in WP: WP Gatsby and WP GraphQL. However, I want to be able to build the site using Gutenberg and ACF blocks. Doing some googling lead me to a the following plugin WP GraphQL Gutenberg v0.3.11. After installing this the development server now long spins up. The Gutenberg plugin hasn't been updated for over 2 years. Is it required or is the block compatibility already baked into one of the other plugins @TylerBarnes? 

Any help would be appreciated. Attempting to building a pro bono charity site for a friend and wanted to use this stack.

**Comment by TylerBarnes** [comment]
I don't recommend using `wp-graphql-gutenberg with` Gatsby as it's schema is quite complex which leads to slow builds. @jasonbahl has mentioned wanting to add Gutenberg support directly to WPGraphQL, but I'm not sure what the latest news on that is. @jasonbahl is also currently doing a lot of work on `wp-graphql-acf` but I'm not sure if that includes support for Gutenberg ACF blocks.
Right now the happy path is definitely using ACF fully w/ flexible content instead of Gutenberg

**Comment by stevepepple** [comment]
Good to know about `wp-graphql-gutenberg` @TylerBarnes. Our site will require those blocks for some time, so will keep an eye out for any updates to the code GraphQL library in this regard. 

In my experience, MediaItems seem to cause the biggest strain during the build process, take the longest, and fail the most. Any pointers for making that step more efficient? Even when `createFileNodes` is false, it still take a long time. We tried to offload to a CDN, but it didn't exactly make the build much smoother.

**Comment by TylerBarnes** [comment]
@stevepepple last week I enabled excluding fields on the MediaItem type during sourcing https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-wordpress/docs/features/media-item-processing.md#speeding-up-mediaitem-fetching-from-wp

**Comment by stevepepple** [comment]
Oh, brilliant! that probably will make it much for us, as we have over 1K images. What version of the plugin does that require?

**Comment by TylerBarnes** [comment]
@stevepepple it's available in `6.24.1` and `7.2.0`

**Comment by stevepepple** [comment]
Great, I'm trying the new version right now: For some reason it still seems way slower than the other fields. Perhaps I didn't configure it correctly, so I'll debug with the `writeQueriesToDisk` option a bit more and then open a separate issue with my findings. It also might be the the node process is slow to report progress. What often happens is that the process seems to hang on the first 1-20 nodes. 

<img width="456" alt="Screen Shot 2022-12-01 at 3 11 03 PM" src="https://user-images.githubusercontent.com/466699/205180332-bf66cc45-be5f-4623-a97b-01b7d870ed00.png">

**Comment by TylerBarnes** [comment]
That does seem really slow. Let me know what you find! It's also worth trying [WPGraphQL smart cache](https://github.com/wp-graphql/wp-graphql-smart-cache) if your host supports it. Object caching seems to work well with `gatsby-source-wordpress` from the few sites I've tried it on. We don't support the network caching feature yet, but I've seen good results so far from object caching.

**Comment by stevepepple** [comment]
Okay, sure, will let you know. I'd been curious if there was a way to cache the GraphQL queries, so this answer that question. We're not on WPEngine, but still should be able to give it a try and hopefully see benefits. Much appreciated!

**Comment by stevepepple** [comment]
Hey @TylerBarnes, I've experimented with a mix of the smart cache plugin and excluded fields in our Post and MediaImage queries, and the performance is perhaps marginally better, but still about 30mins on the source and transform nodes step, 90% on images. Where I good play to document my findings? (I realize it's slightly off topic for this issue.)

I confirmed that there is a slimmed down query for media, like this: 
```
query NODE_LIST_QUERY($first: Int!, $after: String) {
  mediaItems(first: $first, after: $after) {
    nodes {
      altText
      caption
      databaseId
      description
      guid
      id
      link
      mediaItemUrl
      mediaType
      mimeType
      slug
      sourceUrl
      title
      __typename
    }
    pageInfo {
      hasNextPage
      endCursor
    }
  }
}
```

If I run the same query in the Wordpress GraphQL IDE, it says it's not cached: 
```
"This response was not executed at run-time but has been returned from the GraphQL Object Cache"
```
Using the the IDE queries seem to come back quickly, but I suppose it's not a bunch of similudainous requests.

**Comment by TylerBarnes** [comment]
@stevepepple lets take it to a [new discussion](https://github.com/gatsbyjs/gatsby/discussions/new)

**Comment by stevepepple** [comment]
> @stevepepple lets take it to a [new discussion](https://github.com/gatsbyjs/gatsby/discussions/new)

Discussion and details are here: https://github.com/gatsbyjs/gatsby/discussions/37166
Note: I also updated this with some details about our Wordpress deployment on Digital Ocean using Ngnix. (I have a hunch that perhaps there's some additional optimizations in this configuration that could improve concurrency and performance.)
